### ISSUE TITLE
How to wrap an existing logger?

### ISSUE THREAD

**Issue by lamba92** [MicroUtils/kotlin-logging#88]
Say you already have an `org.slf4j.Logger` object and you want to wrap around it the capabilities of `KLoggeg`. Is it possible?

**Comment by oshai** [comment]
I am not sure I understand what you mean. Can you give a concrete example?

**Comment by lamba92** [comment]
I have a `slf4j` logger from a [`Ktor`](ktor.io) [`application`](https://ktor.io/servers/index.html#application-and-applicationenvironment) and I'd like to wrap it.

**Comment by oshai** [comment]
The api to do that is internal (ie: it's not part of the api), you can look for the method of `wrapJLogger` to see the implementation. But if I understand how ktor works since it is using slf4j you can just use kotlin-logging as usual and it will work (ie: let kotlin-logging instantiate the logger for you.

**Comment by lamba92** [comment]
Yeah i digged inside the code and saw it. I have little knowledge on how to use properly those loggers and I wasn't sure  if creating a new logger was a good idea or not.

**Comment by oshai** [comment]
I don't think you can, those classes are internal.
How do you obtain the slf4j logger?

**Comment by corneil** [comment]
You would ideally want to write code as follows:
```kotlin
companion object {
val logger= KotlinLogging.wrap(slf4jLogger)
}
```
I notice there is a method: `mu.internal.KLoggerFactory.wrapJLogger` that does this work we just need to expose it in `KotlinLogging`

**Comment by oshai** [comment]
@corneil yes, you can see I mentioned this method in one of the comments above. The thing that is not clear to me is why do you need that over just using `KotlinLogging.logger {}` which will have a similar effect assuming the logger is of the current class.

**Comment by lamba92** [comment]
How one should behave then? A simple 
```kotlin
val MyContext.klog by lazy { 
    KotlinLogging.logger{} 
}
``` 
Is ok? If so why?

**Comment by oshai** [comment]
how about:
```
private val logger = KotlinLogging.logger {} 
```
in the file, just outside of the class? you can see the example here: https://github.com/MicroUtils/kotlin-logging#getting-started

**Comment by oshai** [comment]
@lamba92 was the issue resolved?

**Comment by lamba92** [comment]
I have not tried it yet but I guess it is! Still, a wrapper would have been better!

**Comment by oshai** [comment]
@lamba92 How about adding the wrap method to the KotlinLogging? also added an extension method:

```kotlin
actual object KotlinLogging {
    /**
     * This method allow defining the logger in a file in the following way:
     * val logger = KotlinLogging.logger {}
     */
    actual fun logger(func: () -> Unit): KLogger = KLoggerFactory.logger(func)

    actual fun logger(name: String): KLogger = KLoggerFactory.logger(name)

    fun logger(underlyingLogger: Logger) = KLoggerFactory.wrapJLogger(underlyingLogger)
}

fun Logger.toKLogger() = KotlinLogging.logger(this)
```
see also PR #90

**Comment by lamba92** [comment]
Oh great that is great :)
One last thing. Say that I recover my logger from a specific context with property `logger`.

One could now create an extension property `klogger` as such:
```kotlin
val MyContext.klogger
    get() = KotlinLogging.logger(logger)
```
It is fine but every time you use it you are creating a new object.

What about adding a `Boolean` value to the function that allows to save the newly created KLogger into a map with the underlying logger as key? You would be now able to check the map if it already been created and fetch it or create a new one, save it and return it!

In any case I guess a lazy would do the same as:
```kotlin
val MyContext.klogger by lazy { 
    KotlinLogging.logger(logger) 
}
```

**Comment by oshai** [comment]
You can do the map solution, but I think adding it to the library is not that great. You could argue the same even without the new functionality but I think that map will make things more complicated and might lead to memory leaks.

**Comment by corneil** [comment]
I like this solution!


[image: --]
Corneil du Plessis
[image: https://]about.me/corneil
<https://about.me/corneil?promo=email_sig>

**Comment by lamba92** [comment]
> You can do the map solution, but I think adding it to the library is not that great. You could argue the same even without the new functionality but I think that map will make things more complicated and might lead to memory leaks.

Oh yeah, that's memory leaks 101 :P Anyway, the wrapper is fine :) Thanks!

**Comment by lamba92** [comment]
Misclick the close button

**Comment by oshai** [comment]
fixed in 1.7.6
